### PR TITLE
Add vtkCleaverImageToUnstructuredGridFilter

### PR DESCRIPTION
**Pull request by aronhelser**
This initial version does something useful. It provides the same parameters and defaults as the ITK filter that also wraps Cleaver, and constructs an unstructured grid output directly from the Cleaver result mesh. 

The test creates a very course image, with two spherical regions with values 1.0 and 2.0 - a labelmap. To transform the labelmap into the indicator functions Cleaver wants, I mimicked an ITK distance filter by combining two VTK distance filters for inside and outside the region. The test checks for the expected number of cells and points in the output, and has a rendered image baseline comparison. Here's the output:
vtkCleaverImageToUnstructuredGridFilterTest

The padding parameter and indicator function toggle still need to be hooked up.

### CODE REVIEW

**Comment by aronhelser** [comment]
Next, working on creating unstructured grid output directly instead of writing a file. Then wiring up parameters.

**Comment by aronhelser** [comment]
@jcfr this filter is minimally functional now. Still need to hook up padding and indicator function flag. We can start working on integration with something else now...

File: vtkCleaverImageToUnstructuredGridFilter.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+=========================================================================*/
+#ifndef vtkCleaverImageToUnstructuredGridFilter_h
+#define vtkCleaverImageToUnstructuredGridFilter_h
+
+#if defined(_WIN32)
+#define VTKCLEAVER_EXPORT __declspec( dllexport )
+#else
+#define VTKCLEAVER_EXPORT
+#endif
```
Comment (jcfr) [inline vtkCleaverImageToUnstructuredGridFilter.h:22]: Remove this and instead include 

```
#include "vtkCleaverImageToUnstructuredGridModule.h" // For export macro
```

File: vtk.module [hunk 2]
```diff
@@ -0,0 +1,32 @@
+NAME
+  VTK::CleaverImageToUnstructuredGrid
```
Comment (jcfr) [inline vtk.module:2]:
```suggestion
  VTK::Cleaver
```

File: vtk.module [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+LIBRARY_NAME
+  vtkCleaverImageToUnstructuredGrid
```
Comment (jcfr) [inline vtk.module:4]:
```suggestion
  vtkCleaver
```

File: vtkCleaverImageToUnstructuredGridFilter.cxx [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+    blur[jj]->SetInputConnection(threshold[jj]->GetOutputPort());
+    blur[jj]->SetStandardDeviation(sigma);
+    // second threshold
+    thresholdDistance[jj]->ReplaceInOn();
+    thresholdDistance[jj]->SetInValue(0.0);
+    thresholdDistance[jj]->SetInputConnection(blur[jj]->GetOutputPort());
+    // distance map
+    distance[jj]->SetInputConnection(thresholdDistance[jj]->GetOutputPort());
+  }
+  // inverse of each other
```
Comment (jcfr) [inline vtkCleaverImageToUnstructuredGridFilter.cxx:85]: explain why

File: vtkCleaverImageToUnstructuredGridFilter.cxx [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+
+      distance[jj]->Update();
+    }
+    imgdiff->SetOperationToSubtract();
+    imgdiff->SetInputConnection(0, distance[0]->GetOutputPort());
+    imgdiff->SetInputConnection(1, distance[1]->GetOutputPort());
+    imgdiff->Update();
+    auto* img = imgdiff->GetOutput();
+
+    int dims[3];
```
Comment (jcfr) [inline vtkCleaverImageToUnstructuredGridFilter.cxx:118]: I suggest to use `vtkIdType`

**Review by jcfr — COMMENTED** [review]
Few minor nitpicks.

Could you also include screenshot of input and output in the pull request description ?

**Comment by aronhelser** [comment]
I'm used to gitlab - there's not a pull request description at the top, or am I missing something?

The test input is generated in the test, and it's a volume, so no easy screen shot - 24 x 24 x 24 with two partial spherical labels with values 1.0 and 2.0. 

Here's the rendered output:
test

File: vtkCleaverImageToUnstructuredGridFilter.h (hunk 1 above)
Comment (aronhelser) [reply vtkCleaverImageToUnstructuredGridFilter.h:22]: OK, I thought we had to include cmake code for this to be available, but I see the vtk module takes care of it.

**Comment by jcfr** [comment]
> there's not a pull request description at the top, or am I missing something?

image

By editing the description you can then provide a complete summary of all changes.

File: vtkCleaverImageToUnstructuredGridFilter.h (hunk 1 above)
Comment (jcfr) [reply vtkCleaverImageToUnstructuredGridFilter.h:22]: > but I see the vtk module takes care of it.

Indeed, with the current approach everything is taking care of and the next step will only be one possible approach for dissemination.

**Comment by aronhelser** [comment]
OK, using the initial comment makes sense. gitlab has a separate description from the comment stream.

File: vtk.module (hunk 2 above)
Comment (aronhelser) [reply vtk.module]: done

File: vtkCleaverImageToUnstructuredGridFilter.cxx (hunk 4 above)
Comment (aronhelser) [reply vtkCleaverImageToUnstructuredGridFilter.cxx]: done

File: vtkCleaverImageToUnstructuredGridFilter.h (hunk 1 above)
Comment (aronhelser) [reply vtkCleaverImageToUnstructuredGridFilter.h]: fixed.

File: vtkCleaverImageToUnstructuredGridFilter.h [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+  vtkGetMacro(Padding, int);
+  ///@}
+
+  ///@{
+  vtkSetMacro(Alpha, double);
+  vtkGetMacro(Alpha, double);
+  ///@}
+
+protected:
+    vtkCleaverImageToUnstructuredGridFilter();
```
[the remark this answers is not in the document]
Comment (aronhelser) [reply vtkCleaverImageToUnstructuredGridFilter.h]: reformatted with clang-format

File: vtkCleaverImageToUnstructuredGridFilter.h [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    ~vtkCleaverImageToUnstructuredGridFilter();
```
[the remark this answers is not in the document]
Comment (aronhelser) [reply vtkCleaverImageToUnstructuredGridFilter.h]: done

File: vtkCleaverImageToUnstructuredGridFilter.cxx [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+        if ((a.y < b.y) && (b.y - a.y) > 1e-9) return true;
+        else if ((a.y > b.y) && (a.y - b.y) > 1e-9) return false;
+        if ((a.z < b.z) && (b.z - a.z) > 1e-9) return true;
+        else if ((a.z > b.z) && (a.z - b.z) > 1e-9) return false;
+        return false;
+      }
+  };
+typedef std::map< const vec3, size_t, vec3_compare > VertMap;
+// given a cleaver TetMesh, fill in the UnstructuredGrid output. Based on
+// TetMesh::writeVtkUnstructuredGrid()
```
[the remark this answers is not in the document]
Comment (aronhelser) [reply vtkCleaverImageToUnstructuredGridFilter.cxx:211]: done

File: vtkCleaverImageToUnstructuredGridFilter.cxx [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+      blur[jj]->Update();
+      auto* img = blur[jj]->GetOutput();
+
+      img->GetScalarRange(range);
+      // get the midpoint of the range
+      auto md = (range[0] + range[1]) / 2.f;
+      vtkLog(INFO, "Md " << md );
+
+      // ITK itkApproximateSignedDistanceMapImageFilter has inside/outside values
+      // for VTK, anything non-zero is considered NOT an object. Threshold again to
```
[the remark this answers is not in the document]
Comment (aronhelser) [reply vtkCleaverImageToUnstructuredGridFilter.cxx]: I added more comments about why its two parallel pipelines, one for calculating distance outside the label, and one for inside. Then they are combined with subtraction.

File: vtkCleaverImageToUnstructuredGridFilter.cxx [hunk 10]
```diff
@@ -0,0 +1,409 @@
...
+  image->GetScalarRange(range);
+  auto minLabel = static_cast<size_t>(range[0]);
+  auto maxLabel = static_cast<size_t>(range[1]);
+  vtkLog(INFO, "Range " << minLabel << " " << maxLabel );
+  vtkNew<vtkImageThreshold> threshold[2];
+  vtkNew<vtkImageGaussianSmooth> blur[2];
+  vtkNew<vtkImageEuclideanDistance> distance[2];
+  vtkNew<vtkImageThreshold> thresholdDistance[2];
+  vtkNew<vtkImageMathematics> imgdiff;
+  for(auto jj = 0; jj < 2; ++jj)
```
[the remark this answers is not in the document]
Comment (aronhelser) [reply vtkCleaverImageToUnstructuredGridFilter.cxx]: changed to idx

**Comment by aronhelser** [comment]
Thanks for the detailed review - keeping me honest when borrowing code from different tests.
I think I've addressed all comments, but I could have missed something. Do you want to take another look @jcfr ?

**Comment by aronhelser** [comment]
Added support for indicator function inputs. Padding support was already added. Features are now at parity with the ITK filter. Here's the second test image:
image

**Pull request merged**